autoload/lsp/utils.vim

```diff
@@ -138,3 +138,14 @@ function! lsp#utils#error(msg) abort
     echom a:msg
     echohl NONE
 endfunction
+
+function! lsp#utils#echo_with_truncation(msg) abort
+    let l:msg = a:msg
+    let l:winwidth = winwidth(0)
+
+    if l:winwidth < strdisplaywidth(l:msg)
+        let l:msg = l:msg[:l:winwidth - 5] . '...'
```

Comment:
As a heads up, if there a multibyte character doing  a truncation this way is unsafe - this splits across bytes but not display characters. See `:help strcharpart()` for an alternative.

Also you're correctly handling the fact that a character can have a display width greater than 1 by using `strdisplaywidth`, however if you have enough of them the truncated message here could still take multiple lines. The only way I found to solve that issue was to drop a character at a time in a loop until the display width is short enough.

See https://github.com/natebosch/vim-lsc/commit/6cbd77f4df7578d1c97e48ba8cd4206794d5a840

Reply:
Lot of code in vim-lsp currently doesn't support mutli-byte. I do need to fix it.

Thanks for the comment. Seems like we should make it easy to share some of these utils :)